packages/side-nav/src/SideNav.tsx

```diff
@@ -1,53 +1,190 @@
-import React from 'react';
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
 import PropTypes from 'prop-types';
+import { transparentize } from 'polished';
 import { css, cx } from '@leafygreen-ui/emotion';
-import { ulStyleOverrides, sideNavWidth } from './styles';
+import { useEventListener } from '@leafygreen-ui/hooks';
+import { HTMLElementProps, IdAllocator, keyMap } from '@leafygreen-ui/lib';
+import { uiColors } from '@leafygreen-ui/palette';
+import CollapseButton from './CollapseButton';
+import { SideNavContext } from './contexts';
+import {
+  sideNavWidth,
+  sideNavCollapsedWidth,
+  transitionDurationMilliseconds,
+  useHover,
+} from './utils';
 
-const navStyles = css`
+const navStyle = css`
+  position: relative;
+  z-index: 0;
+  min-height: 100%;
+  min-width: ${sideNavWidth}px;
   width: ${sideNavWidth}px;
+
+  background-color: ${uiColors.gray.light3};
+  border-right: 1px solid ${uiColors.gray.light2};
+  transition: all ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavStyle = css`
+  position: absolute;
+  min-width: ${sideNavCollapsedWidth}px;
+  width: ${sideNavCollapsedWidth}px;
+  top: 0;
+  bottom: 0;
+`;
+
+const collapsedHoveredNavStyle = css`
+  position: absolute;
+  top: 0;
+  bottom: 0;
+  border-width: 0;
+  box-shadow: 6px 0 10px -8px ${transparentize(0.7, uiColors.black)};
+`;
+
+const collapsedSpacerStyle = css`
+  min-width: ${sideNavCollapsedWidth}px;
+  width: ${sideNavCollapsedWidth}px;
 `;
 
-interface SideNavProps {
-  /**
-   * Class name that will be applied to the root-level element.
-   */
+const navListStyle = css`
+  position: relative;
+  padding-top: 16px;
+  height: 100%;
+  overflow: hidden;
+  transition: all ${transitionDurationMilliseconds}ms ease-in-out;
+`;
+
+const collapsedNavListStyle = css`
+  padding-top: 64px;
+`;
+
+type Props = {
+  children: React.ReactNode;
   className?: string;
+  currentPath?: string;
+} & (
+  | { collapsible: false; initialCollapsed?: undefined }
+  | { collapsible?: true; initialCollapsed?: boolean }
+) &
+  Omit<HTMLElementProps<'nav'>, 'ref'>;
+
+const idAllocator = IdAllocator.create('side-nav');
+
+const SideNav = React.forwardRef<HTMLElement, Props>(function SideNav(
+  {
+    children,
+    className,
+    currentPath,
+    collapsible = true,
+    initialCollapsed = false,
+    onMouseOver: onMouseOverProp,
+    onMouseLeave: onMouseLeaveProp,
+    ...rest
+  }: Props,
+  ref,
+) {
+  const id = useMemo(() => idAllocator.generate(), []);
+
+  // eslint-disable-next-line prefer-const
+  let [collapsed, setCollapsed] = useState(
+    collapsible ? initialCollapsed : false,
+  );
+
+  if (!collapsible) {
+    collapsed = false;
```

Comment:
Why not setCollapsed here

Reply:
(Why do we need this outside of an inside of the useEffect hook)

Reply:
this is so that the following code we can just write `collapsed` instead of `collapsible && collapsed` everywhere and render the correct thing on this render (and not the next render that would be triggered by the useEffect hook)

Reply:
I might be missing something but doesn't the ternary above read if it's collapsible then use the initialCollapsed prop and if it's not collapsible set it to false.

Reply:
we're branching on the same variable in both places

Reply:
the argument to `useState` only applies to the initial render